io/log_stream.go

```diff
@@ -0,0 +1,38 @@
+package io
+
+import (
+	"context"
+	"io"
+)
+
+type logBody struct {
+	ctx    context.Context
+	fmtstr string
+	log    func(format string, args ...interface{})
+	input  io.ReadCloser
+}
+
+func newLogBody(ctx context.Context, fmtstr string, log func(format string, args ...interface{}), rc io.ReadCloser) io.ReadCloser {
+	return &logBody{
+		ctx:    ctx,
+		fmtstr: fmtstr,
+		input:  rc,
+		log:    log,
+	}
+}
+
+func (lb *logBody) Read(p []byte) (int, error) {
+	n, err := lb.input.Read(p)
+	if n > 0 {
+		lb.log("%s%s", lb.fmtstr, p)
+	}
+	return n, err
+}
+
+func (lb *logBody) Close() error {
+	return lb.input.Close()
+}
+
+func LogBody(ctx context.Context, fmtstr string, log func(format string, args ...interface{}), rc io.ReadCloser) io.ReadCloser {
```

Comment:
Lets just pass `func([]byte)` callback and let caller decide what and how to log

Reply:
why? It's such a common interface function for loggers!

Reply:
Maybe you want to count bytes. Also the whole point of format string is to defer evaluation of arguments until logging and not evaluate them if log level is disabled. In your case
```go
		req.Body = io.LogBody(
			req.Context(),
			fmt.Sprintf(`logBody("request") %s: `, req.Header.Get(flowid.HeaderName)),
			ctx.Logger().Infof,
			req.Body,
		)
```
you always spend time creating format string even if there will be no logging (e.g. when log level set to error).
Compare to
```go
		req.Body = io.LogBody(
			req.Context(),
			func(chunk []byte) {
				ctx.Logger().Infof(`logBody("request") %s: %s`, req.Header.Get(flowid.HeaderName), chunk)
			},
			req.Body,
		)
```

Reply:
ok got it

Reply:
changed